source/fuzz/transformation_access_chain.cpp

```diff
@@ -196,21 +280,149 @@ protobufs::Transformation TransformationAccessChain::ToMessage() const {
 }
 
 std::pair<bool, uint32_t> TransformationAccessChain::GetIndexValue(
-    opt::IRContext* ir_context, uint32_t index_id) const {
+    opt::IRContext* ir_context, uint32_t index_id,
+    uint32_t object_type_id) const {
+  if (!ValidIndexToComposite(ir_context, index_id, object_type_id)) {
+    return {false, 0};
+  }
   auto index_instruction = ir_context->get_def_use_mgr()->GetDef(index_id);
-  if (!index_instruction || !spvOpcodeIsConstant(index_instruction->opcode())) {
-    // TODO(https://github.com/KhronosGroup/SPIRV-Tools/issues/3179) We could
-    //  allow non-constant indices when looking up non-structs, using clamping
-    //  to ensure they are in-bounds.
+
+  uint32_t bound = fuzzerutil::GetBoundForCompositeIndex(
+      *ir_context->get_def_use_mgr()->GetDef(object_type_id), ir_context);
+
+  // The index must be a constant
+  if (!spvOpcodeIsConstant(index_instruction->opcode())) {
+    return {false, 0};
+  }
+
+  // The index must be in bounds.
+  uint32_t value = index_instruction->GetSingleWordInOperand(0);
+
+  if (value >= bound) {
+    return {false, 0};
+  }
+
+  return {true, value};
+}
+
+std::pair<bool, uint32_t> TransformationAccessChain::CreateAndGetClampedIndexId(
+    opt::IRContext* ir_context, uint32_t index_id, uint32_t object_type_id,
+    bool add_clamping_instructions,
+    std::pair<uint32_t, uint32_t> fresh_ids) const {
+  if (!ValidIndexToComposite(ir_context, index_id, object_type_id)) {
     return {false, 0};
   }
+
+  // The composite must not be a struct
+  auto object_type_def = ir_context->get_def_use_mgr()->GetDef(object_type_id);
+
+  if (object_type_def->opcode() == SpvOpTypeStruct) {
+    return {false, 0};
+  }
+
+  // Valid fresh ids need to have been given
+  if (fresh_ids.first <= 0 || fresh_ids.second <= 0) {
+    return {false, 0};
+  }
+
+  // Perform the clamping using the fresh ids at our disposal.
+  // The module will not be changed if |add_clamping_instructions| is not set.
+  auto index_instruction = ir_context->get_def_use_mgr()->GetDef(index_id);
+
+  uint32_t bound =
+      fuzzerutil::GetBoundForCompositeIndex(*object_type_def, ir_context);
+
+  if (!TryToClampInteger(ir_context, *index_instruction, bound, fresh_ids,
+                         add_clamping_instructions)) {
+    // It was not possible to clamp the integer
+    return {false, 0};
+  }
+
+  // The clamped integer will be at id |fresh_ids[1]|.
+  return {true, fresh_ids.second};
```

Comment:
The fact that - regardless of the rest of the computation - this function returns something deterministic as the second component in its pair - isn't good.

I suggest (as per the above comment) that you get rid of this function in any case.

But if you were to keep it, it might be better to make it simply return a boolean, and to have the "Apply" code that invokes it know that fresh_ids.second is the result id.

Reply:
Good point. I inlined the function so this is no longer relevant in this specific case.